### PR TITLE
[OpenAPI] case class field being an Option of an ADT are not rendered as nullable

### LINKED ISSUE #3052: [OpenAPI] case class field being an `Option` of an ADT are not rendered as nullable
```scala
sealed trait Example
final case class A(a: Int) extends Example
final case class B(b: String) extends Example

/**
 * This is the Input documentation
 */
final case class Input(example: Option[Example])

val testEndpoint = 
  (Endpoint(RoutePattern.POST / "test") ?? Doc.p("This is my 'POST /test' endpoint doc"))
    .in[Input]
    .out[String](mediaType = MediaType.application.json, doc = Doc.p("this is the output doc"))

val spec: String =
  OpenAPIGen.fromEndpoints(
    title = "This is my OpenAPI doc title",
    version = "0.0.0",
    endpoints = List(testEndpoint)
  ).toJson
```

The generated OpenAPI spec will be:
```json
"Input": {
  "type": "object",
  "properties": {
    "example": {
      "$ref": "#/components/schemas/Example"
    },
  }
}
```
According to https://stackoverflow.com/a/48114924/2431728, it should be:
```json
"Input": {
  "type": "object",
  "properties": {
    "example": [
      "anyOf": [
        { "type": "null" },
        { "$ref": "#/components/schemas/Example" }
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by guizmaii**
Fixes https://github.com/zio/zio-http/issues/3052

### CODE REVIEW

File: zio-http/shared/src/main/scala/zio/http/endpoint/openapi/OpenAPIGen.scala [hunk 1]
```diff
@@ -988,7 +988,7 @@ object OpenAPIGen {
   }
   private def schemaReferencePath(nominal: TypeId.Nominal, referenceType: SchemaStyle): String = {
     referenceType match {
-      case SchemaStyle.Compact => s"#/components/schemas/${nominal.typeName}}"
```
Comment (guizmaii) [inline zio-http/shared/src/main/scala/zio/http/endpoint/openapi/OpenAPIGen.scala:991]: Nasty typo, which was sadly validated by the tests here https://github.com/zio/zio-http/pull/3053/files#diff-0b5e5bc79ca85db4957d0ab782e9509a15f69b70d84ffd76a15a712367a860f3L2234-L2236

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala [hunk 2]
```diff
@@ -2989,6 +2997,114 @@ object OpenAPIGenSpec extends ZIOSpecDefault {
...
+            |        {
+            |        "type" :
+            |          "object",
+            |        "properties" : {
+            |          "optionalAdtField" : {
+            |            "anyOf": [
+            |              { "type": "null" },
+            |              { "$ref": "#/components/schemas/SealedTraitCustomDiscriminator" }
+            |            ]
+            |          }
```
Comment (guizmaii) [inline zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:3068]: We now correctly have the `anyOf`

Comment (987Nabil) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:3068]: I wonder, if this should be anyOf, or if the field should just be missing from the required fields list. I'd tend to the second. Most Scala json serialization does not write the field if the option is None

Comment (987Nabil) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:3068]: Maybe both. To be accept all possible encodings

Comment (guizmaii) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:3069]: I made the change. Now, we have both 🙂

Comment (guizmaii) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:3068]: Done. They are now both nullable and non-required

**Review by jdegoes — APPROVED** [review]

**Pull request merged**